{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Training Wheel Exercise: Renal Cancer Data\n",
    "\n",
    "Objective: \n",
    "Identifying relevant features is a common procedure in biological research (e.g gene expression and protein expression studies). Hence, comprehensive understanding of its workflow is mandatory. \n",
    "\n",
    "The purpose of this exercise is to provide an introduction into the conventional pipeline for pathway enrichment studies. More specifically, protein expression data will be analysed. Here, we are employing commonly-used feature selection methods e.g. t-test, recursive feature elimination, and selecting top proteins based on statistical p-value ranks. \n",
    "\n",
    "Subsequently, the frequently used hypergeometric enrichment tool will be utilised to evaluate enrichment of these protein features to pathways. \n",
    "\n",
    "The results from this exercise can also be used as a benchmark as we perform feature selection through \"fuzzy-logic\" tool\n",
    "s in future studies. \n",
    "\n",
    "Footnote: \n",
    "This notebook also serves as a logbook for BS9001."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Installation of packages \n",
    "\n",
    "Let's first install the packages that are required for analyses. Do ensure that pip is already installed, otherwise, download pip in the command line. \n",
    "\n",
    "Installation is not required if it's already done so. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'python' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-a4637bbefc43>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpython\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mversion\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'python' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pandas in /usr/local/lib/python3.9/site-packages (1.2.3)\n",
      "Requirement already satisfied: numpy>=1.16.5 in /usr/local/lib/python3.9/site-packages (from pandas) (1.20.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.9/site-packages (from pandas) (2.8.1)\n",
      "Requirement already satisfied: pytz>=2017.3 in /usr/local/lib/python3.9/site-packages (from pandas) (2021.1)\n",
      "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.9/site-packages (from python-dateutil>=2.7.3->pandas) (1.15.0)\n",
      "\u001b[33mWARNING: You are using pip version 20.3.3; however, version 21.1.1 is available.\n",
      "You should consider upgrading via the '/usr/local/opt/python@3.9/bin/python3.9 -m pip install --upgrade pip' command.\u001b[0m\n",
      "Requirement already satisfied: sklearn in /usr/local/lib/python3.9/site-packages (0.0)\n",
      "Requirement already satisfied: scikit-learn in /usr/local/lib/python3.9/site-packages (from sklearn) (0.24.1)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.9/site-packages (from scikit-learn->sklearn) (2.1.0)\n",
      "Requirement already satisfied: joblib>=0.11 in /usr/local/lib/python3.9/site-packages (from scikit-learn->sklearn) (1.0.1)\n",
      "Requirement already satisfied: numpy>=1.13.3 in /usr/local/lib/python3.9/site-packages (from scikit-learn->sklearn) (1.20.1)\n",
      "Requirement already satisfied: scipy>=0.19.1 in /usr/local/lib/python3.9/site-packages (from scikit-learn->sklearn) (1.6.1)\n",
      "\u001b[33mWARNING: You are using pip version 20.3.3; however, version 21.1.1 is available.\n",
      "You should consider upgrading via the '/usr/local/opt/python@3.9/bin/python3.9 -m pip install --upgrade pip' command.\u001b[0m\n",
      "Requirement already satisfied: matplotlib in /usr/local/lib/python3.9/site-packages (3.3.4)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in /usr/local/lib/python3.9/site-packages (from matplotlib) (2.8.1)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.9/site-packages (from matplotlib) (1.3.1)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.9/site-packages (from matplotlib) (8.1.2)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.3 in /usr/local/lib/python3.9/site-packages (from matplotlib) (2.4.7)\n",
      "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.9/site-packages (from matplotlib) (0.10.0)\n",
      "Requirement already satisfied: numpy>=1.15 in /usr/local/lib/python3.9/site-packages (from matplotlib) (1.20.1)\n",
      "Requirement already satisfied: six in /usr/local/lib/python3.9/site-packages (from cycler>=0.10->matplotlib) (1.15.0)\n",
      "\u001b[33mWARNING: You are using pip version 20.3.3; however, version 21.1.1 is available.\n",
      "You should consider upgrading via the '/usr/local/opt/python@3.9/bin/python3.9 -m pip install --upgrade pip' command.\u001b[0m\n",
      "Requirement already satisfied: numpy in /usr/local/lib/python3.9/site-packages (1.20.1)\n",
      "\u001b[33mWARNING: You are using pip version 20.3.3; however, version 21.1.1 is available.\n",
      "You should consider upgrading via the '/usr/local/opt/python@3.9/bin/python3.9 -m pip install --upgrade pip' command.\u001b[0m\n",
      "Requirement already satisfied: bioinfokit in /usr/local/lib/python3.9/site-packages (2.0.4)\n",
      "Requirement already satisfied: tabulate in /usr/local/lib/python3.9/site-packages (from bioinfokit) (0.8.9)\n",
      "Requirement already satisfied: scikit-learn in /usr/local/lib/python3.9/site-packages (from bioinfokit) (0.24.1)\n",
      "Requirement already satisfied: pandas in /usr/local/lib/python3.9/site-packages (from bioinfokit) (1.2.3)\n",
      "Requirement already satisfied: scipy in /usr/local/lib/python3.9/site-packages (from bioinfokit) (1.6.1)\n",
      "Requirement already satisfied: matplotlib-venn in /usr/local/lib/python3.9/site-packages (from bioinfokit) (0.11.6)\n",
      "Requirement already satisfied: adjustText in /usr/local/lib/python3.9/site-packages (from bioinfokit) (0.7.3)\n",
      "Requirement already satisfied: seaborn in /usr/local/lib/python3.9/site-packages (from bioinfokit) (0.11.1)\n",
      "Requirement already satisfied: textwrap3 in /usr/local/lib/python3.9/site-packages (from bioinfokit) (0.9.2)\n",
      "Requirement already satisfied: numpy in /usr/local/lib/python3.9/site-packages (from bioinfokit) (1.20.1)\n",
      "Requirement already satisfied: statsmodels in /usr/local/lib/python3.9/site-packages (from bioinfokit) (0.12.2)\n",
      "Requirement already satisfied: matplotlib in /usr/local/lib/python3.9/site-packages (from bioinfokit) (3.3.4)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.9/site-packages (from matplotlib->bioinfokit) (8.1.2)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.9/site-packages (from matplotlib->bioinfokit) (1.3.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.9/site-packages (from matplotlib->bioinfokit) (0.10.0)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.3 in /usr/local/lib/python3.9/site-packages (from matplotlib->bioinfokit) (2.4.7)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in /usr/local/lib/python3.9/site-packages (from matplotlib->bioinfokit) (2.8.1)\n",
      "Requirement already satisfied: six in /usr/local/lib/python3.9/site-packages (from cycler>=0.10->matplotlib->bioinfokit) (1.15.0)\n",
      "Requirement already satisfied: pytz>=2017.3 in /usr/local/lib/python3.9/site-packages (from pandas->bioinfokit) (2021.1)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.9/site-packages (from scikit-learn->bioinfokit) (2.1.0)\n",
      "Requirement already satisfied: joblib>=0.11 in /usr/local/lib/python3.9/site-packages (from scikit-learn->bioinfokit) (1.0.1)\n",
      "Requirement already satisfied: patsy>=0.5 in /usr/local/lib/python3.9/site-packages (from statsmodels->bioinfokit) (0.5.1)\n",
      "\u001b[33mWARNING: You are using pip version 20.3.3; however, version 21.1.1 is available.\n",
      "You should consider upgrading via the '/usr/local/opt/python@3.9/bin/python3.9 -m pip install --upgrade pip' command.\u001b[0m\n",
      "Collecting combat\n",
      "  Using cached combat-0.3.0-py3-none-any.whl (36 kB)\n",
      "Requirement already satisfied: patsy==0.5.1 in /usr/local/lib/python3.9/site-packages (from combat) (0.5.1)\n",
      "Collecting mpmath==1.1.0\n",
      "  Using cached mpmath-1.1.0.tar.gz (512 kB)\n",
      "Collecting numpy==1.18.5\n",
      "  Using cached numpy-1.18.5.zip (5.4 MB)\n",
      "  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h    Preparing wheel metadata ... \u001b[?25lerror\n",
      "\u001b[31m    ERROR: Command errored out with exit status 1:\n",
      "     command: /usr/local/opt/python@3.9/bin/python3.9 /usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py prepare_metadata_for_build_wheel /var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/tmp_kq27i00\n",
      "         cwd: /private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3\n",
      "    Complete output (209 lines):\n",
      "    Running from numpy source directory.\n",
      "    setup.py:461: UserWarning: Unrecognized setuptools command, proceeding with generating Cython sources and expanding templates\n",
      "      run_build = parse_setuppy_commands()\n",
      "    Processing numpy/random/_bounded_integers.pxd.in\n",
      "    Processing numpy/random/_philox.pyx\n",
      "    Processing numpy/random/_bit_generator.pyx\n",
      "    Processing numpy/random/_bounded_integers.pyx.in\n",
      "    Processing numpy/random/_sfc64.pyx\n",
      "    Processing numpy/random/_mt19937.pyx\n",
      "    Processing numpy/random/mtrand.pyx\n",
      "    Processing numpy/random/_generator.pyx\n",
      "    Processing numpy/random/_pcg64.pyx\n",
      "    Processing numpy/random/_common.pyx\n",
      "    Cythonizing sources\n",
      "    blas_opt_info:\n",
      "    blas_mkl_info:\n",
      "    customize UnixCCompiler\n",
      "      libraries mkl_rt not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    blis_info:\n",
      "      libraries blis not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    openblas_info:\n",
      "      libraries openblas not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    atlas_3_10_blas_threads_info:\n",
      "    Setting PTATLAS=ATLAS\n",
      "      libraries tatlas not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    atlas_3_10_blas_info:\n",
      "      libraries satlas not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    atlas_blas_threads_info:\n",
      "    Setting PTATLAS=ATLAS\n",
      "      libraries ptf77blas,ptcblas,atlas not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    atlas_blas_info:\n",
      "      libraries f77blas,cblas,atlas not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    accelerate_info:\n",
      "      libraries accelerate not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "    Library accelerate was not found. Ignoring\n",
      "      libraries veclib not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "    Library veclib was not found. Ignoring\n",
      "      FOUND:\n",
      "        extra_compile_args = ['-msse3', '-I/System/Library/Frameworks/vecLib.framework/Headers']\n",
      "        extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']\n",
      "        define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]\n",
      "    \n",
      "      FOUND:\n",
      "        extra_compile_args = ['-msse3', '-I/System/Library/Frameworks/vecLib.framework/Headers']\n",
      "        extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']\n",
      "        define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]\n",
      "    \n",
      "    non-existing path in 'numpy/distutils': 'site.cfg'\n",
      "    lapack_opt_info:\n",
      "    lapack_mkl_info:\n",
      "      libraries mkl_rt not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    openblas_lapack_info:\n",
      "      libraries openblas not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    openblas_clapack_info:\n",
      "      libraries openblas,lapack not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    flame_info:\n",
      "      libraries flame not found in ['/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib', '/usr/local/lib', '/usr/lib']\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    atlas_3_10_threads_info:\n",
      "    Setting PTATLAS=ATLAS\n",
      "      libraries lapack_atlas not found in /usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib\n",
      "      libraries tatlas,tatlas not found in /usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib\n",
      "      libraries lapack_atlas not found in /usr/local/lib\n",
      "      libraries tatlas,tatlas not found in /usr/local/lib\n",
      "      libraries lapack_atlas not found in /usr/lib\n",
      "      libraries tatlas,tatlas not found in /usr/lib\n",
      "    <class 'numpy.distutils.system_info.atlas_3_10_threads_info'>\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    atlas_3_10_info:\n",
      "      libraries lapack_atlas not found in /usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib\n",
      "      libraries satlas,satlas not found in /usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib\n",
      "      libraries lapack_atlas not found in /usr/local/lib\n",
      "      libraries satlas,satlas not found in /usr/local/lib\n",
      "      libraries lapack_atlas not found in /usr/lib\n",
      "      libraries satlas,satlas not found in /usr/lib\n",
      "    <class 'numpy.distutils.system_info.atlas_3_10_info'>\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    atlas_threads_info:\n",
      "    Setting PTATLAS=ATLAS\n",
      "      libraries lapack_atlas not found in /usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib\n",
      "      libraries ptf77blas,ptcblas,atlas not found in /usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib\n",
      "      libraries lapack_atlas not found in /usr/local/lib\n",
      "      libraries ptf77blas,ptcblas,atlas not found in /usr/local/lib\n",
      "      libraries lapack_atlas not found in /usr/lib\n",
      "      libraries ptf77blas,ptcblas,atlas not found in /usr/lib\n",
      "    <class 'numpy.distutils.system_info.atlas_threads_info'>\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "    atlas_info:\n",
      "      libraries lapack_atlas not found in /usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib\n",
      "      libraries f77blas,cblas,atlas not found in /usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib\n",
      "      libraries lapack_atlas not found in /usr/local/lib\n",
      "      libraries f77blas,cblas,atlas not found in /usr/local/lib\n",
      "      libraries lapack_atlas not found in /usr/lib\n",
      "      libraries f77blas,cblas,atlas not found in /usr/lib\n",
      "    <class 'numpy.distutils.system_info.atlas_info'>\n",
      "      NOT AVAILABLE\n",
      "    \n",
      "      FOUND:\n",
      "        extra_compile_args = ['-msse3', '-I/System/Library/Frameworks/vecLib.framework/Headers']\n",
      "        extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']\n",
      "        define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]\n",
      "    \n",
      "    /usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/dist.py:274: UserWarning: Unknown distribution option: 'define_macros'\n",
      "      warnings.warn(msg)\n",
      "    running dist_info\n",
      "    running build_src\n",
      "    build_src\n",
      "    building py_modules sources\n",
      "    creating build\n",
      "    creating build/src.macosx-11-x86_64-3.9\n",
      "    creating build/src.macosx-11-x86_64-3.9/numpy\n",
      "    creating build/src.macosx-11-x86_64-3.9/numpy/distutils\n",
      "    building library \"npymath\" sources\n",
      "    Traceback (most recent call last):\n",
      "      File \"/usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py\", line 280, in <module>\n",
      "        main()\n",
      "      File \"/usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py\", line 263, in main\n",
      "        json_out['return_val'] = hook(**hook_input['kwargs'])\n",
      "      File \"/usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py\", line 133, in prepare_metadata_for_build_wheel\n",
      "        return hook(metadata_directory, config_settings)\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-build-env-ti46lajw/overlay/lib/python3.9/site-packages/setuptools/build_meta.py\", line 166, in prepare_metadata_for_build_wheel\n",
      "        self.run_setup()\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-build-env-ti46lajw/overlay/lib/python3.9/site-packages/setuptools/build_meta.py\", line 258, in run_setup\n",
      "        super(_BuildMetaLegacyBackend,\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-build-env-ti46lajw/overlay/lib/python3.9/site-packages/setuptools/build_meta.py\", line 150, in run_setup\n",
      "        exec(compile(code, __file__, 'exec'), locals())\n",
      "      File \"setup.py\", line 488, in <module>\n",
      "        setup_package()\n",
      "      File \"setup.py\", line 480, in setup_package\n",
      "        setup(**metadata)\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/core.py\", line 171, in setup\n",
      "        return old_setup(**new_attr)\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-build-env-ti46lajw/overlay/lib/python3.9/site-packages/setuptools/__init__.py\", line 153, in setup\n",
      "        return distutils.core.setup(**attrs)\n",
      "      File \"/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/core.py\", line 148, in setup\n",
      "        dist.run_commands()\n",
      "      File \"/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/dist.py\", line 966, in run_commands\n",
      "        self.run_command(cmd)\n",
      "      File \"/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/dist.py\", line 985, in run_command\n",
      "        cmd_obj.run()\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-build-env-ti46lajw/overlay/lib/python3.9/site-packages/setuptools/command/dist_info.py\", line 31, in run\n",
      "        egg_info.run()\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/command/egg_info.py\", line 26, in run\n",
      "        self.run_command(\"build_src\")\n",
      "      File \"/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/cmd.py\", line 313, in run_command\n",
      "        self.distribution.run_command(command)\n",
      "      File \"/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/dist.py\", line 985, in run_command\n",
      "        cmd_obj.run()\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/command/build_src.py\", line 146, in run\n",
      "        self.build_sources()\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/command/build_src.py\", line 157, in build_sources\n",
      "        self.build_library_sources(*libname_info)\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/command/build_src.py\", line 290, in build_library_sources\n",
      "        sources = self.generate_sources(sources, (lib_name, build_info))\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/command/build_src.py\", line 380, in generate_sources\n",
      "        source = func(extension, build_dir)\n",
      "      File \"numpy/core/setup.py\", line 656, in get_mathlib_info\n",
      "        st = config_cmd.try_link('int main(void) { return 0;}')\n",
      "      File \"/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/command/config.py\", line 241, in try_link\n",
      "        self._check_compiler()\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/command/config.py\", line 82, in _check_compiler\n",
      "        self.fcompiler = new_fcompiler(compiler=self.fcompiler,\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/fcompiler/__init__.py\", line 884, in new_fcompiler\n",
      "        compiler = get_default_fcompiler(plat, requiref90=requiref90,\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/fcompiler/__init__.py\", line 855, in get_default_fcompiler\n",
      "        compiler_type =  _find_existing_fcompiler(matching_compiler_types,\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/fcompiler/__init__.py\", line 806, in _find_existing_fcompiler\n",
      "        c.customize(dist)\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/fcompiler/__init__.py\", line 530, in customize\n",
      "        linker_so_flags = self.flag_vars.linker_so\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/fcompiler/environment.py\", line 40, in __getattr__\n",
      "        return self._get_var(name, conf_desc)\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/fcompiler/environment.py\", line 56, in _get_var\n",
      "        var = self._hook_handler(name, hook)\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/fcompiler/__init__.py\", line 709, in _environment_hook\n",
      "        return hook()\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/fcompiler/gnu.py\", line 353, in get_flags_linker_so\n",
      "        flags = GnuFCompiler.get_flags_linker_so(self)\n",
      "      File \"/private/var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/pip-install-ard82_ss/numpy_7bd293f9244342a08a3c58be1ae3b9e3/numpy/distutils/fcompiler/gnu.py\", line 144, in get_flags_linker_so\n",
      "        os.environ['MACOSX_DEPLOYMENT_TARGET'] = target\n",
      "      File \"/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/os.py\", line 684, in __setitem__\n",
      "        value = self.encodevalue(value)\n",
      "      File \"/usr/local/Cellar/python@3.9/3.9.1_3/Frameworks/Python.framework/Versions/3.9/lib/python3.9/os.py\", line 756, in encode\n",
      "        raise TypeError(\"str expected, not %s\" % type(value).__name__)\n",
      "    TypeError: str expected, not int\n",
      "    ----------------------------------------\u001b[0m\n",
      "\u001b[31mERROR: Command errored out with exit status 1: /usr/local/opt/python@3.9/bin/python3.9 /usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py prepare_metadata_for_build_wheel /var/folders/q8/b50ytlzd33bdkl879p0p7klm0000gp/T/tmp_kq27i00 Check the logs for full command output.\u001b[0m\n",
      "\u001b[33mWARNING: You are using pip version 20.3.3; however, version 21.1.1 is available.\n",
      "You should consider upgrading via the '/usr/local/opt/python@3.9/bin/python3.9 -m pip install --upgrade pip' command.\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?25hRequirement already satisfied: seaborn in /usr/local/lib/python3.9/site-packages (0.11.1)\n",
      "Requirement already satisfied: numpy>=1.15 in /usr/local/lib/python3.9/site-packages (from seaborn) (1.20.1)\n",
      "Requirement already satisfied: scipy>=1.0 in /usr/local/lib/python3.9/site-packages (from seaborn) (1.6.1)\n",
      "Requirement already satisfied: matplotlib>=2.2 in /usr/local/lib/python3.9/site-packages (from seaborn) (3.3.4)\n",
      "Requirement already satisfied: pandas>=0.23 in /usr/local/lib/python3.9/site-packages (from seaborn) (1.2.3)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.3 in /usr/local/lib/python3.9/site-packages (from matplotlib>=2.2->seaborn) (2.4.7)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.9/site-packages (from matplotlib>=2.2->seaborn) (1.3.1)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.9/site-packages (from matplotlib>=2.2->seaborn) (8.1.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.9/site-packages (from matplotlib>=2.2->seaborn) (0.10.0)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in /usr/local/lib/python3.9/site-packages (from matplotlib>=2.2->seaborn) (2.8.1)\n",
      "Requirement already satisfied: six in /usr/local/lib/python3.9/site-packages (from cycler>=0.10->matplotlib>=2.2->seaborn) (1.15.0)\n",
      "Requirement already satisfied: pytz>=2017.3 in /usr/local/lib/python3.9/site-packages (from pandas>=0.23->seaborn) (2021.1)\n",
      "\u001b[33mWARNING: You are using pip version 20.3.3; however, version 21.1.1 is available.\n",
      "You should consider upgrading via the '/usr/local/opt/python@3.9/bin/python3.9 -m pip install --upgrade pip' command.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "#to install the libraries and packages required \n",
    "#before this run, download pip in command line \n",
    "!pip install pandas\n",
    "!pip install sklearn \n",
    "!pip install matplotlib\n",
    "!pip install numpy\n",
    "!pip install bioinfokit\n",
    "!pip install combat\n",
    "!pip install seaborn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data preparation \n",
    "In this exercise, the well-studied renal cancer (RC) data will be used for analyses.\n",
    "\n",
    "Some of the characteristics of the RC data include the following: \n",
    "- Protein expression data \n",
    "- \"Cleaned\" data \n",
    "- Two groups of patients: normal (control) and cancer (case) group\n",
    "- Consist of 6 patients in each group with data obtained in duplicates \n",
    "- Consist of 3123 protein expression genes \n",
    "- Dimensions: 3123 rows x 24 columns  \n",
    "\n",
    "Firstly, data transformation needs to be performed in order to render the RC data appropriate for subsequent analyses. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unnamed: 0              Q9UBE0  Q9BSJ8  P02656 O95741  P09651  P55809  Q15631  \\\n",
      "cancer_cc_patient1_rep1  40914   41185  274731   4969  103836  101089  102971   \n",
      "cancer_cc_patient1_rep2  45120   42150  284693   5472  118185   97593  110008   \n",
      "cancer_cc_patient2_rep1  44113  113386  141656   7872  162475  137794  112840   \n",
      "cancer_cc_patient2_rep2  47835  139305  155864   7957  170174   99304  129621   \n",
      "cancer_cc_patient3_rep1  26957   35891  161075   6111  101960   88960   92826   \n",
      "cancer_cc_patient3_rep2  26005   30788  124223   3722   82866   68415   75856   \n",
      "cancer_cc_patient6_rep1  35712   52750  194500   4094  118675  117222  142954   \n",
      "cancer_cc_patient6_rep2  30517   33753  166061   3580   83659   85197   92312   \n",
      "cancer_cc_patient7_rep1  38094   62466  177344   5878   95034   96055   97456   \n",
      "cancer_cc_patient7_rep2  26654   45894  133566   4066   71776   54583   58171   \n",
      "cancer_cc_patient8_rep1  48576  113950  121262   5700  175963  122041  114483   \n",
      "cancer_cc_patient8_rep2  29078   80761   86180   5558  118167   84064   78399   \n",
      "normal_cc_patient1_rep1  24671   51134  231306  18527  116804  124316   91593   \n",
      "normal_cc_patient1_rep2  27205   60310  246950  20128  118739  132116  113144   \n",
      "normal_cc_patient2_rep1  33659   67804  117811   6239  113987  144444  100269   \n",
      "normal_cc_patient2_rep2  31811   70406  130393   6994  108445  146314  106587   \n",
      "normal_cc_patient3_rep1  27149   49577  276229  14445  104072  127128   98140   \n",
      "normal_cc_patient3_rep2  28459   57563  281124  16152  110674  135511  109197   \n",
      "normal_cc_patient6_rep1  24325   51237  187400   7824   99804  159276   95440   \n",
      "normal_cc_patient6_rep2  18261   37385  133036   5350   68386  114049   63663   \n",
      "normal_cc_patient7_rep1  27651   59370  153722  13301  110596  126153   92551   \n",
      "normal_cc_patient7_rep2  19498   43370  110053   9097   62897   89809   64331   \n",
      "normal_cc_patient8_rep1  28037   43835  191212   6608   82070  248441   86715   \n",
      "normal_cc_patient8_rep2  21038   38577  145489   5087   58367  193519   54443   \n",
      "\n",
      "Unnamed: 0               Q96EY1  Q16706 Q9UID3  ...  P50395  P21283   Q08257  \\\n",
      "cancer_cc_patient1_rep1   92750   66723   1011  ...  375745  104892   468298   \n",
      "cancer_cc_patient1_rep2   93422   74885   1241  ...  397595  126349   486071   \n",
      "cancer_cc_patient2_rep1   79844   99927   3380  ...  680440   63593   678865   \n",
      "cancer_cc_patient2_rep2   81800  105634   4129  ...  617383   67945   723119   \n",
      "cancer_cc_patient3_rep1   55847  125118   1474  ...  581510   89942  1115051   \n",
      "cancer_cc_patient3_rep2   45622   73982   1773  ...  431137  134710   850757   \n",
      "cancer_cc_patient6_rep1   62290  101223   2288  ...  537714   90381  1331752   \n",
      "cancer_cc_patient6_rep2   51818   61013   2296  ...  401509   79514  1060288   \n",
      "cancer_cc_patient7_rep1   52474  103429   2254  ...  431058   64136   989674   \n",
      "cancer_cc_patient7_rep2   45658   66908   2729  ...  376628   54144   780264   \n",
      "cancer_cc_patient8_rep1  127651  129296   2918  ...  593276  109032   595285   \n",
      "cancer_cc_patient8_rep2   91185   49525   4957  ...  478246   65525   462075   \n",
      "normal_cc_patient1_rep1   89540   46112   1314  ...  515406  133530  1253792   \n",
      "normal_cc_patient1_rep2   94479   51348   1794  ...  526770  146422  1312312   \n",
      "normal_cc_patient2_rep1   95158   28220   2785  ...  499955  134779   929974   \n",
      "normal_cc_patient2_rep2  101829   32349   2871  ...  504784  143479   995347   \n",
      "normal_cc_patient3_rep1   94452   25677   1783  ...  479795  130177  1080790   \n",
      "normal_cc_patient3_rep2   95789   31524   1794  ...  472938  138071  1190681   \n",
      "normal_cc_patient6_rep1   97324   23813   2022  ...  483650  125869  1509303   \n",
      "normal_cc_patient6_rep2   68180   13736   1515  ...  360582   90036  1204142   \n",
      "normal_cc_patient7_rep1   93435   24067   3131  ...  527874  148985  1184047   \n",
      "normal_cc_patient7_rep2   58510   16609   2226  ...  438460   87200   915990   \n",
      "normal_cc_patient8_rep1   84499   33725   2296  ...  489978  119112  1155422   \n",
      "normal_cc_patient8_rep2   67848   26140   1643  ...  369664   91050   925706   \n",
      "\n",
      "Unnamed: 0               P51161  Q9H2A2  P13010  O95182  P62241 Q14573  Q99436  \n",
      "cancer_cc_patient1_rep1  160322   42805  219862   67041   75779  19285   57186  \n",
      "cancer_cc_patient1_rep2  165311   44656  217378   72618   86028   9108   53835  \n",
      "cancer_cc_patient2_rep1   65667   43342  382306  107180  198237  12929   92997  \n",
      "cancer_cc_patient2_rep2   64751   45178  417187  104060  185249  13245  101256  \n",
      "cancer_cc_patient3_rep1   47259   34210  279052   77867   81408   7177   71378  \n",
      "cancer_cc_patient3_rep2   33815   25084  245144   54533   56156   5195   48542  \n",
      "cancer_cc_patient6_rep1   52249   36464  338998   72188  105342   8554   84505  \n",
      "cancer_cc_patient6_rep2   41425   27075  264573   57676   79724   6544   63219  \n",
      "cancer_cc_patient7_rep1   56073   34713  289712   98885   82905   9602   85674  \n",
      "cancer_cc_patient7_rep2   43675   26344  229266   50533   64655   5935   61742  \n",
      "cancer_cc_patient8_rep1  109079   64418  425587  116255  251496  10294   89637  \n",
      "cancer_cc_patient8_rep2   82732   44439  312771   74321  174284   8627   61633  \n",
      "normal_cc_patient1_rep1   46491  116947  258363  170200   69267  10662   55202  \n",
      "normal_cc_patient1_rep2   51424  122311  281541  177604   67568  10543   70656  \n",
      "normal_cc_patient2_rep1   51527   80248  243869  193690  104207  13154   77741  \n",
      "normal_cc_patient2_rep2   52377   84190  246101  196868  108275  14506   92645  \n",
      "normal_cc_patient3_rep1   55383   96298  224645  199734   66510  12986   73448  \n",
      "normal_cc_patient3_rep2   54171  105925  242725  226574   65616  13730   93582  \n",
      "normal_cc_patient6_rep1   46776  129391  243594  173373   86740  13042   84393  \n",
      "normal_cc_patient6_rep2   36792   85152  165345  133652   54269   8373   46821  \n",
      "normal_cc_patient7_rep1   49098   96324  292678  173675   92693  12787   88170  \n",
      "normal_cc_patient7_rep2   35183   57736  207920  126919   62146   8265   48752  \n",
      "normal_cc_patient8_rep1   43522  108274  287338  215409   64100  10473   67560  \n",
      "normal_cc_patient8_rep2   33929   73382  208732  155882   46373   7711   55797  \n",
      "\n",
      "[24 rows x 3123 columns]\n"
     ]
    }
   ],
   "source": [
    "#Data preparation \n",
    "%matplotlib inline\n",
    "import pandas as pd \n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt \n",
    "from __future__ import division\n",
    "from sklearn.decomposition import PCA \n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn import preprocessing\n",
    "\n",
    "#transpose and convert first column as header, generate the feature tables\n",
    "geneaggregate = pd.read_csv('RC_data.csv')\n",
    "df = geneaggregate[sorted(geneaggregate.columns)]\n",
    "dfT = df.swapaxes(\"index\", \"columns\")\n",
    "new_header = dfT.iloc[0] #grab the first row for the header\n",
    "featuretable = dfT[1:] #take the data less the header row\n",
    "featuretable.columns = new_header #set the header row as the df header\n",
    "\n",
    "#store the protein expression values and protein gene names \n",
    "samples = featuretable.values\n",
    "sample_names = featuretable.columns\n",
    "\n",
    "print(featuretable)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Batch Correction \n",
    "Biological data are often obtained in batches. Technical sources of variation across batches, can lead to heterogenity across batches of data (e.g. different experiment personnels, different experimental instruments) - also known as batch effects.\n",
    "\n",
    "Batch effects can potentially confound data, and lead to discrepancies in statistical testing. Such discrepancies can obscure revelation of important explanatory variables in datasets (e.g. subpopulations). \n",
    "\n",
    "Hence, batch correction is often an imperative step to ensure that discoveries made from the data in question are truly meaningful and relevant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-a061e6734c35>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m#batch correction\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m#transform data into suitable format for batch correction\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mdf2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Unnamed: 0\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mdf2_logtransformed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlog2\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df' is not defined"
     ]
    }
   ],
   "source": [
    "from combat import pycombat\n",
    "#batch correction \n",
    "#transform data into suitable format for batch correction \n",
    "df2 = df.set_index(\"Unnamed: 0\")\n",
    "df2_logtransformed = np.log2(df2)\n",
    "\n",
    "#perform batch correction\n",
    "batch = [1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2]\n",
    "df_batchcorrected = pycombat(df2_logtransformed,batch)\n",
    "\n",
    "#plt.boxplot(df_corrected.transpose())\n",
    "#plt.show()\n",
    "df_batchcorrected_T = df_batchcorrected.T\n",
    "df_batchcorrected_T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
